### PR TITLE
[interactive_media_ads] Adds initial Android implementation

### PR DESCRIPTION
**Pull request by bparrishMines**
Android implementation for https://github.com/flutter/flutter/issues/134228

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/interactive_media_ads/android/build.gradle
Comment (bparrishMines) [inline packages/interactive_media_ads/android/build.gradle:63]: This is the same line added to `camera_android_camerax`:  https://github.com/flutter/packages/blob/main/packages/camera/camera_android_camerax/android/build.gradle#L77

File: packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/ProxyApiRegistrar.kt
Comment (bparrishMines) [inline packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/ProxyApiRegistrar.kt:19]: This was added to the registrar to make it easier to handle and test callback methods on the correct thread.

File: packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/VideoAdPlayerProxyApi.kt
Comment (bparrishMines) [inline packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/VideoAdPlayerProxyApi.kt:66]: Custom methods added since `VideoAdPlayer.getVolume` and `VideoAdPlayer.getAdProgress` require a synchronous callback method.

File: packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart
Comment (bparrishMines) [inline packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart:161]: I started a new style of setting callback methods, which can cause a circular reference, in a static method. I would forget the rule of not using non-local methods in callback methods and this help prevented me from accidentally calling a method/field in the encapsulating class.

File: packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/AdErrorProxyApi.kt
Comment (stuartmorgan) [inline packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/AdErrorProxyApi.kt:36]: It looks like this doesn't handle the case where a new error code is added upstream. It seems like it would probably even fail to compile if that happened?

File: packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/BaseDisplayContainerProxyApi.kt
Comment (stuartmorgan) [inline packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/BaseDisplayContainerProxyApi.kt:7]: Shouldn't all of the classes have class-level comments? https://developer.android.com/kotlin/style-guide#documentation

File: packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/VideoAdPlayerProxyApi.kt
Comment (stuartmorgan) [reply packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/VideoAdPlayerProxyApi.kt:66]: I think it's worth adding this as a code comment to these methods. It would be good to have it clarify why it's not calling `api.foo` in addition to saving the value.

File: packages/interactive_media_ads/lib/interactive_media_ads.dart
Comment (stuartmorgan) [inline packages/interactive_media_ads/lib/interactive_media_ads.dart:11]: In some cases in the past we've regretted directly re-exposing platform interface APIs in the app-facing API (instead of having a second version and converting); have you audited these to make sure we're very unlikely to ever want to make any breaking changes to them, such as adding a required constructor param, or in having some method that implementations need but clients should never use?

File: packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart
Comment (stuartmorgan) [inline packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart:62]: Could you explain in the comment where/how this number was derived? I.e., is it arbitrary, based on best practice docs, other?

File: packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart
Comment (stuartmorgan) [inline packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart:68]: Could you elaborate on the comments here and/or for `_videoAdPlayer`? After reading both comments it's not really clear to me what the responsibilities of each are.

File: packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart
Comment (stuartmorgan) [inline packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart:80]: Maybe explain the difference between a display container and a view (`_videoView`) here?

File: packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart
Comment (stuartmorgan) [inline packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart:120]: This could use a comment about intended use, since it only resets a subset of the state.

File: packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart
Comment (stuartmorgan) [reply packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart:153]: Maybe add this reference to the method comment?

File: packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart
Comment (stuartmorgan) [reply packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart:161]: It's probably worth briefly commenting this pattern in the code each time it's used so that future people understand the benefits and aren't tempted to "simplify" the code by removing this (or at least not replicating it in new code).

File: packages/interactive_media_ads/lib/src/android/interactive_media_ads_proxy.dart
Comment (stuartmorgan) [inline packages/interactive_media_ads/lib/src/android/interactive_media_ads_proxy.dart:52]: Wouldn't some of these be more readable without the trailing comma? Or would the inconsistency hurt readability?

File: packages/interactive_media_ads/pigeons/interactive_media_ads_android.dart
Comment (stuartmorgan) [inline packages/interactive_media_ads/pigeons/interactive_media_ads_android.dart:5]: Why is everything in this file commented out?

**Review by stuartmorgan — COMMENTED** [review]
(I only briefly skimmed the tests

File: packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/GeneratedInteractiveMediaAdsLibrary.kt
Comment (bparrishMines) [reply packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/GeneratedInteractiveMediaAdsLibrary.kt:1]: Good catch. I forgot that we only did that because of the limitations of Java and Obj-c.

File: packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/AdErrorProxyApi.kt
Comment (bparrishMines) [reply packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/AdErrorProxyApi.kt:42]: My intention was for this to break compilation when a new enum is added. Since this uses a pinned version of the IMA SDK, we can add a new enum to the code generation whenever we bump the version.

We add an `unknown` enum in the `webview_flutter` implementations, but I assumed we did that because it uses a system library and not a pinned SDK.

File: packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart
Comment (bparrishMines) [reply packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart:75]: It was the same value as the one in the android example: https://developers.google.com/interactive-media-ads/docs/sdks/android/client-side#6.-create-the-videoadplayeradapter-class. It doesn't give an explanation, but I added a comment to explain where the value is from.

File: packages/interactive_media_ads/lib/src/android/interactive_media_ads_proxy.dart
Comment (bparrishMines) [reply packages/interactive_media_ads/lib/src/android/interactive_media_ads_proxy.dart]: Yea, it does make it more readable. I removed the trailing commas

File: packages/interactive_media_ads/pigeons/interactive_media_ads_android.dart
Comment (bparrishMines) [reply packages/interactive_media_ads/pigeons/interactive_media_ads_android.dart:5]: This uses features from https://github.com/flutter/packages/pull/6371, so it would break tests without it commented out. But I'll go ahead and uncomment it for now so it can be reviewed while I still wait for https://github.com/flutter/packages/pull/6371 to land.

File: packages/interactive_media_ads/lib/interactive_media_ads.dart
Comment (bparrishMines) [reply packages/interactive_media_ads/lib/interactive_media_ads.dart:11]: Does the fact that the platform interface, the implementations, and the app-facing API live in the same package affect this? This should ideally give us more flexibility to make private methods/constructors/subclasses if we need to.

File: packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/AdErrorProxyApi.kt
Comment (stuartmorgan) [reply packages/interactive_media_ads/android/src/main/kotlin/dev/flutter/packages/interactive_media_ads/AdErrorProxyApi.kt:42]: > Since this uses a pinned version of the IMA SDK

It doesn't really though, because of [optimistic upgrades](https://docs.gradle.org/current/userguide/resolution_strategy_tuning.html#fail-version-conflict). Because of the way Gradle works, unless we want to make the entire build fail on conflicts (which would be a large hammer) we don't actually have any control over whether we get the version listed in build.gradle.

File: packages/interactive_media_ads/lib/interactive_media_ads.dart
Comment (stuartmorgan) [reply packages/interactive_media_ads/lib/interactive_media_ads.dart:11]: I guess we could always move these classes to the app-facing directory and make a new class that the platform interface code used, if we need to make breaking changes later, since changing the platform interface wouldn't be public in this structure.

**Review by stuartmorgan — APPROVED** [review]
LGTM once the remaining comment about the enum fallback is resolved.

**Comment by reidbaker** [comment]
This is on my review queue. Out of curiosity will the android team maintain the android implementation of this plugin like we do for some others or will this be owned by the ecosystem team?

**Comment by stuartmorgan** [comment]
For now it'll be ecosystem; medium term we can discuss. Longer term, the goal is to hand off to the SDK owners.

**Comment by bparrishMines** [comment]
@reidbaker I'm submitting this now since I have the LGTM from @stuartmorgan. But still feel free to review or skim when you are able to.

**Comment by reidbaker** [comment]
Can you give me another day?

**Comment by bparrishMines** [comment]
Ah, sorry. The autosubmit bot merged it quick after. The plugin is still in alpha/beta, so we can still make breaking changes after this publishes.

**Pull request merged**
